File: src/main/java/itmo/is/project/service/user/AccountService.java
package itmo.is.project.service.user;

import itmo.is.project.dto.user.AccountDto;
import itmo.is.project.mapper.user.AccountMapper;
import itmo.is.project.model.user.Account;
import itmo.is.project.model.user.User;
import itmo.is.project.repository.user.AccountRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AccountService {
    private final AccountRepository accountRepository;
    private final AccountMapper accountMapper;

    @Transactional
    public void createAccount(User user) {
        accountRepository.findByUserId(user.getId()).ifPresent(account -> {
            throw new IllegalStateException("Account already exists");
        });
        Account account = new Account();
        account.setUser(user);
        accountRepository.save(account);
    }

    public Page<AccountDto> findAllAccounts(Pageable pageable) {
        return accountRepository.findAll(pageable).map(accountMapper::toDto);
    }

    public AccountDto findAccountByUserId(Integer userId) {
        return accountMapper.toDto(findByUserId(userId));
    }

    @Transactional
    public AccountDto deposit(Integer userId, Integer amount) {
        Account account = findByUserId(userId);
        account.deposit(amount);
        account = accountRepository.save(account);
        return accountMapper.toDto(account);
    }

    @Transactional
    public AccountDto withdraw(Integer userId, Integer amount) {
        Account account = findByUserId(userId);
        account.withdraw(amount);
        if (account.getBalance() < 0) {
            throw new IllegalStateException("Insufficient funds");
        }
        account = accountRepository.save(account);
        return accountMapper.toDto(account);
    }

    @Transactional
    public void transferFundsBetweenStationAndUser(Integer userId, int stationBalanceChange) {
        Account user = findByUserId(userId);
        Account station = accountRepository.findOwnerAccount()
                .orElseThrow(() -> new EntityNotFoundException("Owner account not found"));
        station.setBalance(station.getBalance() + stationBalanceChange);
        user.setBalance(user.getBalance() - stationBalanceChange);
        if (station.getBalance() < 0 || user.getBalance() < 0) {
            throw new IllegalStateException("Insufficient funds");
        }
        accountRepository.save(user);
        accountRepository.save(station);
    }

    private Account findByUserId(Integer userId) {
        return accountRepository.findByUserId(userId)
                .orElseThrow(() -> new EntityNotFoundException("Account not found with user id: " + userId));
    }
}

File: src/main/kotlin/itmo/isproject/service/user/AccountService.kt
package itmo.isproject.service.user

import itmo.isproject.dto.user.AccountDto
import itmo.isproject.mapper.user.AccountMapper
import itmo.isproject.model.user.Account
import itmo.isproject.model.user.User
import itmo.isproject.repository.user.AccountRepository
import jakarta.persistence.EntityNotFoundException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AccountService(
    private val accountRepository: AccountRepository,
    private val accountMapper: AccountMapper
) {

    @Transactional
    fun createAccount(user: User) {
        accountRepository.findByUserId(user.id)?.also { _ ->
            throw IllegalStateException("Account already exists")
        }
        val account = Account()
        account.user = user
        accountRepository.save(account)
    }

    fun findAllAccounts(pageable: Pageable): Page<AccountDto> {
        return accountRepository.findAll(pageable).map { accountMapper.toDto(it) }
    }

    fun findAccountByUserId(userId: Int): AccountDto {
        return accountMapper.toDto(findByUserId(userId))
    }

    @Transactional
    fun deposit(userId: Int, amount: Int): AccountDto {
        var account = findByUserId(userId)
        account.deposit(amount)
        account = accountRepository.save(account)
        return accountMapper.toDto(account)
    }

    @Transactional
    fun withdraw(userId: Int, amount: Int): AccountDto {
        var account = findByUserId(userId)
        account.withdraw(amount)
        if (account.balance < 0) {
            throw IllegalStateException("Insufficient funds")
        }
        account = accountRepository.save(account)
        return accountMapper.toDto(account)
    }

    @Transactional
    fun transferFundsBetweenStationAndUser(userId: Int?, stationBalanceChange: Int) {
        val user = findByUserId(userId)
        val station = accountRepository.findOwnerAccount()
            ?: throw EntityNotFoundException("Owner account not found")
        station.balance += stationBalanceChange
        user.balance -= stationBalanceChange
        if (station.balance < 0 || (user.balance) < 0) {
            throw IllegalStateException("Insufficient funds")
        }
        accountRepository.save(user)
        accountRepository.save(station)
    }

    private fun findByUserId(userId: Int?): Account {
        return accountRepository.findByUserId(userId)
            ?: throw EntityNotFoundException("Account not found with user id: $userId")
    }
}
